ceph/ceph
# rgw/pubsub: add S3 compliant API to master zone

## Purpose
This is for allowing notifications directly from the master zone without need for a dedicated pubsub zone, and multisite.

## Refactoring
```rgw_sync_module_pubsub_rest.cc/h``` was refactored into 3 files:
* ```rgw_sync_module_pubsub_rest.cc/h``` - only the Ceph specific REST API, supported on pubsub zone only
* ```rgw_rest_pubsub.cc/h``` - the S3 compliant pubsub REST API
* ```rgw_rest_pubsub_common.cc/h``` - common code between the above 2

As a small cleanup task, the ``rgw_user(const std::string&)`` ctor was made explicit, as this had the potential of causing issues with the xml encoding functions, and overall a good thing :-)

## New API
Master zone support only S3 compliant API, to avoid conflicts with reserved bucket names, used on the pubsube zone with Ceph specific API.
AWS compliant topic creation API was added, based on: https://docs.aws.amazon.com/sns/latest/api/API_CreateTopic.html

## Tests
### Teuthology
pubsub tests are passing:
http://pulpito.front.sepia.ceph.com/yuvalif-2019-08-21_13:03:41-rgw:multisite-wip-yuval-pubsub-api2-distro-basic-smithi/
> notes: 
> * the multisite failures are regression issues, unrelated to this change
> * notification push tests are skipped, due to infrastructure issues (see: https://github.com/ceph/teuthology/pull/1292)

### Manual Testing
* install rabbitmq server. e.g. ``sudo dnf install rabbitmq-server``
* not necessary to start/enable it, since the binary is executed as part of the tests. note that if there issues with running the rabbitmq-server, the amqp tests will be skipped
* to execute the multisite tests on you machine you would need some python libraries:
```
pip install nose
pip install xmltodict
pip install boto3
pip install boto
pip install pika
```
* assuming all commands are executed from inside the ``build`` directory
* make sure that ceph is not running on the machine where you are executing the tests. e.g. 
```
../src/mstop.sh c1 && ../src/mstop.sh c2 && ../src/stop.sh
```
* copy the following configuration file for the multisite tests, into a file (e.g. ``test_multi.conf``):
```
[DEFAULT]
num_zonegroups = 1
num_zones = 1
gateways_per_zone = 1
log_file = test-multi.log
log_level = 1
file_log_level = 1
num_ps_zones = 1
checkpoint_retries = 60
checkpoint_delay = 5
reconfigure_delay = 5
no_bootstrap = false
```
* due to teuthology issues, the tests that push notifications to an external endpoint (http/amqp) are disabled. to enable them, edit: ``src/test/rgw/rgw_multi/tests_ps.py`` and change ``skip_push_tests`` (at line 31) to ``False``
* to execute the pubsusb (including notifications) tests, run the following from inside the ``build`` dir:
```
MON=1 OSD=1 MGR=0 MSD=0 RGW_MULTI_TEST_CONF=./test_multi.conf nosetests -s ../src/test/rgw/test_multi.py -m "test_ps_*"
```
* when running for the first time, the ``no_bootstrap`` parameter in ``test_multi.conf`` should be set to ``false``. Before any consequent run, you should change to ``true``
* if you want to change the setup in the conf file, you should first stop ceph, change the conf file, and change the ``no_bootstrap`` parameter back to ``false``
* to disable all tests directed to the ``pubsub`` zone, set ``num_ps_zones`` to zero
